docs/design-docs/benchmarks.md

```diff
@@ -0,0 +1,262 @@
+
+**Good Usability**
+* Provide a unified entrance to launch benchmarks.
+* Unify the output format for all the micro-benchmarks and E2E-benchmarks, contains return_code, metrics, raw-output, etc.
+
+## Overall System Design
+
+![Structure of Benchmark Package](../assets/benchmark-structure.png)
+
+The structure of `benchmarks` package can be divided into layers from the bottom up:
+1. Abstract base classes for all kind of benchmarks, including `Benchmark`, `ModelBenchmark`, `Microbenchmark` and `DockerBenchmark`.
+   1. `Benchmark` is the base class for all benchmarks. It defines common interfaces such as `run()`, `_preprocess()`, `_postprocess()`, `_benchmark()`, `add_parser_arguments()` and so on.
+   2. `ModelBenchmark` is the base class for all E2E models. It defines the abstract interfaces that need to be implemented by the subclasses using different frameworks, such as `PytorchBase`, `TFBase` and `ONNXBase`. Each subclass will realize part of the abstract interfaces that is common for models, such as `_init_distributed_setting()`, `_init_dataloader()`, `_create_optimizer()`.
+   3. `Microbenchmark` is the base class for all micro benchmarks. It defines the abstract interfaces that need to be implemented by the subclasses, such as `_process_raw_result()`, `_process_numeric_result()`.
+   4. `DockerBenchmark` is the base class for real workloads based on docker. It also defines the abstract interfaces that need to be implemented by the subclasses.
+2. Derived classes for all implemented benchmarks, realized all the abstract interfaces. The benchmarks will be registered into `BenchmarkRegistry`.
+3. `BenchmarkRegistry` provides a way of benchmark registration, maintains all the registered benchmarks, and supports benchmark launching by `BenchmarkContext`.
+4. `BenchmarkContext` provides the context to launch one benchmark, including name, parameters, platform(CPU, GPU, etc.), and framework(Pytorch, TF, ONNX, etc.).
+5. `BenchmarkResult` defines the structured results for each benchmark in json format, including name, return_code, start_time, end_time, raw_data, summarized metrics, etc.
+
+The `Executor` on the uppermost layer is the entrance for all the benchmarks. It launches the benchmark by `BenchmarkRegistry` and fetch `BenchmarkResult`.
+
+## Detailed Component Design
+
+This chapter will describe the design details of all the components in `benchmarks` package.
+
+### E2E Model Benchmarks
+
+The E2E model benchmarks have 4-layer inheritance relationship.
+
+#### Training 
+
+The general process of model training is:
+
+init_distributed_setting -> generate_dataset -> init_dataloader -> create_model -> create_optimizer -> train
+
+These functions will be executed according to the order in the following figure. The functions that exist in derived class but not in base class are abstract functions.
+
+![Training process and the responsibility for function implementation of every layer](../assets/model-training-process.png)
```

Comment:
Function Execution Order in Training Process

Reply:
Done